### PR TITLE
[export] Add custom op guards

### PR DESCRIPTION
**Pull request by angelayi**
For custom ops that do not have a meta kernel, draft export automatically creates a meta kernel based on the tracing example inputs. To ensure that these assumptions made during tracing is clear to the user, we add assertions into the traced exported program:

An example graph:
```
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, a: "f32[s0, s1]", b: "f32[s2, s3]"):
             # File: /data/users/angelayi/pytorch/test/export/test_draft_export.py:172 in forward, code: res1 = torch.ops.mylib.foo4(a, b)
            _assert_tensor_metadata = torch.ops.aten._assert_tensor_metadata(a, dtype = torch.float32, device = device(type='cpu'));  _assert_tensor_metadata = None
            _assert_tensor_metadata_1 = torch.ops.aten._assert_tensor_metadata(b, dtype = torch.float32, device = device(type='cpu'));  _assert_tensor_metadata_1 = None
            foo4: "f32[u2, u3]" = torch.ops.mylib.foo4.default(a, b);  a = b = None
            return (foo4,)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141285
* __->__ #141072
* #141071

Differential Revision: [D66321129](https://our.internmc.facebook.com/intern/diff/D66321129)

### CODE REVIEW

File: torch/_export/passes/insert_custom_op_guards.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    """
+    This is used by draft_export to insert guards in front of calls to custom
+    operators which have a generated fake kernel.
+    """
+    for node in gm.graph.nodes:
+        if node.op == "call_function" and str(node.target) in ops_to_guard:
+            with _set_node_metadata_hook(
+                gm,
+                functools.partial(
+                    _node_metadata_hook, stack_trace=node.meta["stack_trace"]
```
Comment (pianpwk) [inline torch/_export/passes/insert_custom_op_guards.py:21]: should we say `nn_module_stack = node.meta['nn_module_stack']`? Is draft export x unflattener gonna happen often?

File: torch/_export/passes/insert_custom_op_guards.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                            torch.ops.aten._assert_tensor_metadata.default,
+                            args=(arg,),
+                            kwargs={
+                                "dtype": val.dtype,
+                                "device": val.device,
+                                "layout": val.layout,
+                            },
+                        )
+
+                for kwarg in node.kwargs:
```
Comment (pianpwk) [inline torch/_export/passes/insert_custom_op_guards.py:39]: Don't you need to do `node.kwargs.values()`? Also nit: combine the two loops maybe?

File: torch/_export/passes/insert_custom_op_guards.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                    if isinstance(kwarg, torch.fx.Node) and isinstance(
+                        kwarg.meta.get("val"), torch.Tensor
+                    ):
+                        val = kwarg.meta["val"]
+                        gm.graph.call_function(
+                            torch.ops.aten._assert_tensor_metadata.default,
```
Comment (pianpwk) [inline torch/_export/passes/insert_custom_op_guards.py:45]: this is cool

**Review by pianpwk — APPROVED** [review]
just a couple points

File: torch/_export/passes/insert_custom_op_guards.py (hunk 1 above)
Comment (pianpwk) [reply torch/_export/passes/insert_custom_op_guards.py:21]: oh sorry i can't read

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66321129).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66321129).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66321129).

**Comment by zou3519** [comment]
I want to iron out the interaction between serializing the draft export profile for future use and the custom op guards.

When we serialize the draft export profile for future use, how does someone reuse it? I'm thinking in the ideal case, if someone sees the operator again (during a non-draft export run or something), then they are able to use the custom op again. However, if that's the case, then it is important for us to also produce metadata guards in that situation for safety.

**Pull request closed without merging**